sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/NumberConverter.scala

```diff
@@ -49,12 +49,14 @@ object NumberConverter {
    */
   private def encode(radix: Int, fromPos: Int, value: Array[Byte]): Long = {
     var v: Long = 0L
-    val bound = java.lang.Long.divideUnsigned(-1 - radix, radix) // Possible overflow once
+    // bound will always positive since radix >= 2
+    val bound = java.lang.Long.divideUnsigned(-1 - radix, radix)
     var i = fromPos
     while (i < value.length && value(i) >= 0) {
-      if (v >= bound) {
```

Comment:
nit: seems we can simplify the code
```
if (v < 0) return -1
if (v >= bound) {
  ...
}
```

However, this block of code is really hard to understand and we probably need to add some code comments.
1. why `v < 0` means overflow?
2. why we need to check overflow again? Isn't `v >= bound` sufficient to prove overflow?

Reply:
1. Because v*radix >= v*2  (as radix >= 2), so if v is negative, which mean its bit presentation is start with 1, so if we multiply it by 2, it will cause overflow. 
2. v >= bound is sufficient to v*radix + radix will cause overflow. But we need to check that v*radix + value(i) is overflow or not and v*radix + value(i) <= v*radix + radix

Reply:
we can simplify the code as 
```
if (v < 0) return -1
if (v >= bound) {
   ...
} 
```
but I think its logic is not as clear as the current code

Reply:
ok now I got it, thanks for the explanation! I think it's better to simplify the code which makes it easier to write code comments. The 2 conditions of overflow are very different.

Reply:
I think the root of this misunderstanding problem is we treat a long value as an unsigned long value. I think we should introduce a new class of unsigned long value to provide the needed function such as comparing, check for overflow...

Reply:
Yea that would be better. But for this fix that needs to backport, I think changing less code and adding more code comments is safer.

Reply:
Okay, I've refactor code for changing less original code and adding more comments. Can you re-review? @cloud-fan